### DISCUSSION TITLE
Evaluation expression in debugger

### DISCUSSION THREAD

**Discussion by nano-o** [tlaplus/vscode-tlaplus#279] [Q&A] [answered]
Hello, I'm trying the TLA+ debugger using the nightly version of the extension. I set a breakpoint and the debugger stops there. Next I'd like to evaluate an expression in the current state. Is this supported?

**Comment by lemmy** [comment, marked as answer]
Unfortunately, SANY lacks incremental parsing, preventing the debugger from supporting ad-hoc (debug) expressions at runtime.  While it would be possible to hack something into SANY/TLC, I found adding predefined expressions as "Watch" expressions a good-enough workaround.  Any constant-, state-, and action-level formula can be added as a "Watch".  The screenshot below shows the `TraceLog`, `TraceAccepted`, `TraceAlias`, and `TraceView` expressions evaluated as "Watch" expressions.

<img width="1446" alt="Screenshot 2023-04-06 at 10 09 16 AM" src="https://user-images.githubusercontent.com/88777/230448414-8ee5e9c2-84bb-4a56-83cc-890544f0f2bc.png">

**Comment by lemmy** [comment]
[![IMAGE ALT TEXT](http://img.youtube.com/vi/r7kGbjcSaKU/0.jpg)](http://www.youtube.com/watch?v=r7kGbjcSaKU "TLA+ Debugger: Watch expressions")

**Comment by nano-o** [comment]
Thanks! That's helpful.

**Reply by lemmy** [reply]
Today's nightly build comes with (preliminary) support for actual breakpoint expressions: https://youtu.be/rgS0NtoYZZs
